File: greenbank-spigot/src/main/java/org/flameyosflow/greenbank/GreenBankMain.java
package org.flameyosflow.greenbank;

import java.io.IOException;
import java.io.InputStream;
import java.util.Objects;

import dev.dejvokep.boostedyaml.YamlDocument;

import org.bukkit.scheduler.BukkitRunnable;

import org.flameyosflow.greenbank.api.config.DatabaseConfig;
import org.flameyosflow.greenbank.api.config.MessagesConfig;
import org.flameyosflow.greenbank.api.config.Config;
import org.flameyosflow.greenbank.commands.*;
import org.flameyosflow.greenbank.commands.noteconomy.*;
import org.flameyosflow.greenbank.database.mongodb.MongoDBDatabaseConnect;
import org.flameyosflow.greenbank.economy.VaultEconomySupport;
import org.flameyosflow.greenbank.economy.VaultLayer;
import org.flameyosflow.greenbank.listeners.JoinListener;

public class GreenBankMain extends GreenBankSupport {
    private static GreenBankMain instance;

    private VaultEconomySupport economy;

    private MongoDBDatabaseConnect mongoConnect;

    private YamlDocument configFile;

    private YamlDocument databaseConfig;

    private YamlDocument messagesConfig;


    @Override
    public void onLoad() {
        try {
            configFile = new Config(this).configFile();
            databaseConfig = new DatabaseConfig(this).databaseConfig();
            messagesConfig = new MessagesConfig(this).messagesConfig();
        } catch (IOException error) {
            error.printStackTrace();
        }
    }

    @Override
    public void onEnable() {
        setInstance(this);

        if (!setupEconomy()) {
            fireCriticalError("Can't enable Greenbank the economy provider due to vault not being found.", "Please download vault:");
        }

        // if MongoDB exists in db.yml, use that.
        mongoConnect = new MongoDBDatabaseConnect(this);
        BukkitRunnable connectDatabaseAsync = new BukkitRunnable() {
            @Override
            public void run() {
                mongoConnect.connect();
            }
        };

        if (configFile.getBoolean("should-use-async-connection")) {
            connectDatabaseAsync.runTaskAsynchronously(this);
        } else {
            connectDatabaseAsync.runTask(this);
        }

        getLogger().info(" ");
        getLogger().info("Synchronous connection to MongoDB database successfully complete.");

        registerCommand("balance", new BalanceCommand(this));
        registerCommand("pay", new PayCommand(this));
        registerCommand("eco", new EcoCommand(this));
        registerCommand("greenbank", new GreenbankCommand(this));

        registerListener(new JoinListener(this), this);
    }

    @Override
    public void onDisable() {
        try {
            initConfig();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }

        getLogger().info("[GreenBank] GreenBank has been successfully disabled!");
    }

    public void initConfig() throws IOException {
        databaseConfig.save();
        messagesConfig.save();
        configFile.save();

        databaseConfig.reload();
        messagesConfig.reload();
        configFile.reload();
        getLogger().info("Successfully saved and reloaded ALL configuration files.");
    }

    public void logError(String error) {
        getLogger().severe(() -> error);
    }

    public InputStream getConfigYaml(String fileName) {
        try {
            return Objects.requireNonNull(getResource(fileName));
        } catch (NullPointerException error) {
            error.printStackTrace();
            return null;
        }
    }

    public void fireCriticalError(String message, String error) {
        logError("---CRITICAL ERROR!---");
        logError(" ");
        logError(message);
        logError(" ");
        logError(error + " Disabling Greenbank...");
        logError("If you think this is a bug");
        logError("or glitch please contact the");
        logError("owner or staff of this plugin.");
        logError("---------------------");

        // Do not save players or money, just shutdown
        getServer().getPluginManager().disablePlugin(this);
    }

    private boolean setupEconomy() {
        if (getServer().getPluginManager().getPlugin("Vault") == null) {
            return false;
        }

        economy = new VaultEconomySupport(this);
        VaultLayer vaultHook = new VaultLayer(this);
        vaultHook.hook();

        getLogger().info("Economy initialized and registered successfully.");
        return true;
    }

    private void setInstance(GreenBankMain plugin) {
        instance = plugin;
    }

    public static GreenBankMain getPlugin() {
        return instance;
    }

    public YamlDocument getConfigFile() {
        return configFile;
    }

    public YamlDocument getDatabaseConfigFile() {
        return databaseConfig;
    }

    public YamlDocument getMessagesConfigFile() {
        return messagesConfig;
    }

    public VaultEconomySupport getEconomy() {
        return economy;
    }

    public MongoDBDatabaseConnect getMongoConnect() {
        return mongoConnect;
    }
}

File: greenbank-spigot/src/main/java/org/flameyosflow/greenbank/GreenBankMain.kt
package org.flameyosflow.greenbank

import dev.dejvokep.boostedyaml.YamlDocument

import net.milkbowl.vault.economy.Economy

import org.bukkit.Bukkit
import org.bukkit.plugin.ServicePriority
import org.bukkit.scheduler.BukkitRunnable
import org.flameyosflow.greenbank.api.config.Config
import org.flameyosflow.greenbank.api.config.DatabaseConfig
import org.flameyosflow.greenbank.api.config.MessagesConfig
import org.flameyosflow.greenbank.api.economy.VaultEconomySupport
import org.flameyosflow.greenbank.database.mongodb.MongoDBDatabaseConnect

import java.io.IOException; import java.io.InputStream; import java.util.*

class GreenBankMain : GreenBankSupport() {
    var instance: GreenBankMain? = null
    var economy: VaultEconomySupport? = null
    var mongoConnect: MongoDBDatabaseConnect? = null
    var configFile: YamlDocument? = null
    var databaseConfigFile: YamlDocument? = null
    var messagesConfigFile: YamlDocument? = null
    var isShutdown = false

    override fun onLoad() {
        try {
            configFile = Config(this).configFile()
            databaseConfigFile = DatabaseConfig(this).databaseConfig()
            messagesConfigFile = MessagesConfig(this).messagesConfig()
        } catch (error: IOException) { error.printStackTrace() }
        economy = VaultEconomySupport(this); mongoConnect = MongoDBDatabaseConnect(this)
    }

    override fun onEnable() {
        if (!setupEconomy()) fireCriticalError("Cannot find vault on your server, please download vault for functionality", "Please download vault:")
        if (mongoConnect!!.connection != null) {
            val connectDatabaseAsync: BukkitRunnable = object : BukkitRunnable() { override fun run() { mongoConnect!!.createConnection() } }
            when {
                databaseConfigFile!!.getBoolean("should-use-async-connection") -> connectDatabaseAsync.runTaskAsynchronously(this)
                !databaseConfigFile!!.getBoolean("should-use-async-connection") -> connectDatabaseAsync.runTask(this)
            }
        } else { fireCriticalError("Incorrect MongoDB Credentials", "Make sure your MongoDB URI, Database and Collection are correct in db.yml") }
        registerAll(this)
    }

    override fun onDisable(): Unit = try { initConfig() } catch (e: IOException) { throw RuntimeException(e) } finally { logger.info("[GreenBank] GreenBank has been successfully disabled!") }

    private fun fireCriticalError(message: String, error: String?) {
        logger.severe("---CRITICAL ERROR!---")
        logger.severe(message)
        logger.severe(" ")
        logger.severe("$error: Disabling Greenbank...")
        logger.severe("If you think this is a bug or glitch please contact the owner or staff of this plugin.")
        logger.severe("---------------------")
        isShutdown = true; server.pluginManager.disablePlugin(this)
    }

    @Throws(IOException::class) fun initConfig() {
        databaseConfigFile!!.save(); messagesConfigFile!!.save(); configFile!!.save()
        databaseConfigFile!!.reload(); messagesConfigFile!!.reload(); configFile!!.reload()
    }

    fun getConfigYaml(fileName: String?): InputStream = getResource(fileName!!) ?: throw NullPointerException("Filename of getConfigYaml() cannot be null")

    private fun setupEconomy(): Boolean {
        if (server.pluginManager.getPlugin("Vault") == null) return false
        Bukkit.getServicesManager().register(Economy::class.java, economy!!, this, ServicePriority.Highest)
        logger.info("Economy initialized and registered successfully.")
        return true
    }
}